src/display/editor/stamp.js

```diff
@@ -0,0 +1,411 @@
+   * renew its dimensions, hence it must redraw its own contents.
+   * @param {number} width - the new width of the div
+   * @param {number} height - the new height of the div
+   * @returns
+   */
+  #setDimensions(width, height) {
+    height = this.getHeight(width, height);
+    const [parentWidth, parentHeight] = this.parentDimensions;
+    this.width = width / parentWidth;
+    this.height = height / parentHeight;
+    this.setDims(width, height);
+    if (this.#resizeTimeoutId !== null) {
+      clearTimeout(this.#resizeTimeoutId);
+    }
+    this.#resizeTimeoutId = setTimeout(() => {
+      this.#resizeTimeoutId = null;
+      this.#drawBitmap(width, height);
+    }, 200);
+  }
+
+  #scaleBitmap(width, height) {
+    const { width: bitmapWidth, height: bitmapHeight } = this.#bitmap;
+
+    let newWidth = bitmapWidth;
+    let newHeight = bitmapHeight;
+    let bitmap = this.#bitmap;
+    while (newWidth > 2 * width || newHeight > 2 * height) {
+      const prevWidth = newWidth;
+      const prevHeight = newHeight;
+
+      if (newWidth > 2 * width) {
+        // See bug 1820511 (Windows specific bug).
+        // TODO: once the above bug is fixed we could revert to:
+        // newWidth = Math.ceil(newWidth / 2);
+        newWidth =
+          newWidth >= 16384
+            ? Math.floor(newWidth / 2) - 1
+            : Math.ceil(newWidth / 2);
+      }
+      if (newHeight > 2 * height) {
+        newHeight =
+          newHeight >= 16384
+            ? Math.floor(newHeight / 2) - 1
+            : Math.ceil(newHeight / 2);
+      }
+
+      const offscreenCanvas = new OffscreenCanvas(newWidth, newHeight);
+      const offscreenCtx = offscreenCanvas.getContext("2d");
+      offscreenCtx.drawImage(
+        bitmap,
+        0,
+        0,
+        prevWidth,
+        prevHeight,
+        0,
+        0,
+        newWidth,
+        newHeight
+      );
+      bitmap = offscreenCanvas.transferToImageBitmap();
+    }
+
+    return bitmap;
+  }
+
+  #drawBitmap(width, height) {
+    const canvas = this.#canvas;
+    if (!canvas || (canvas.width === width && canvas.height === height)) {
+      return;
+    }
+    canvas.width = width;
+    canvas.height = height;
+    const bitmap = this.#scaleBitmap(width, height);
+    const ctx = canvas.getContext("2d");
+    ctx.filter = this._uiManager.hcmFilter;
+    ctx.drawImage(
+      bitmap,
+      0,
+      0,
+      bitmap.width,
+      bitmap.height,
+      0,
+      0,
+      width,
+      height
+    );
+  }
+
+  #serializeBitmap(toUrl) {
+    if (toUrl) {
+      // We convert to a data url because it's sync and the url can live in the
+      // clipboard.
+      const canvas = document.createElement("canvas");
+      canvas.width = this.#bitmap.width;
+      canvas.height = this.#bitmap.height;
+      const ctx = canvas.getContext("2d");
+      const { width, height } = this.#bitmap;
+      canvas.width = width;
+      canvas.height = height;
```

Comment:
Wait, aren't we setting the canvas width/height to the same values *twice* here?
```suggestion
      const ctx = canvas.getContext("2d");
      const { width, height } = this.#bitmap;
      canvas.width = width;
      canvas.height = height;
```